pyteal/ast/opup.py

```diff
@@ -0,0 +1,141 @@
+                ...,
+                opup.execute(Int(1000))
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            assert target_app_id != None
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCreate:
+            self.target_app_id = Global.current_application_id + Int(1)
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+
+        self.mode = mode
+
+    def _create_app(self) -> Expr:
+        create_app_expr = Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: NoOpApp,
+                    TxnField.clear_state_program: NoOpApp,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+        if self.mode == OpUpMode.OnCall:
+            self.target_app_id = self.target_app_id_slot.load()
+            return Seq(
+                create_app_expr,
+                self.target_app_id_slot.store(
+                    Gitxn[0].created_application_id
+                ),
+            )
+
+        return create_app_expr
+
+    def _begin(self) -> Expr:
+        return If(Global.current_application_id == Int(0)).Then(
+            self._create_app()
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _end(self) -> Expr:
+        return If(Txn.on_completion() == OnComplete.DeleteApplication).Then(
+            self._delete_app(self.target_app_id)
+        )
+
+    # setup() should only be used in OnCreate mode
+    def setup(self) -> Expr:
+        assert(self.mode == OpUpMode.OnCreate)
+        return Seq(
+            self._begin(),
+            self._end(),
+        )
+
+    def _opup_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def execute(self, required_budget: Expr) -> Expr:
+        if self.mode == OpUpMode.OnCall:
+            return If(Global.opcode_budget() < required_budget).Then(
```

Comment:
Have you considered what happens around boundary conditions? There are two cases I can think of:
* The program already has extremely low budget (e.g. 1 unit left) before entering this function. Likely it will fail because you can't even check if you need to op up successfully with such a small budget, so you should put in the docs what the minimum op budget should be before calling the function.
* The program has only slightly more than the amount it asked for. There _might_ be a situation where this function thinks it has enough budget, but the act of returning from the function causes the budget to drop a few units below the amount the caller asked for (due to branch statements that still need to get executed). I've not verified this is the case, but if you know X additional budget is needed to return from the function, you could check `If(Global.opcode_budget() < required_budget + Int(X))`.

Reply:
Another way to phrase the above questions is: how much budget is used before the first inner txn to op up, and how much budget is used after the last `Global.opcode_budget()` check?

Reply:
Good point. A call to `execute` should guarantee that the opcode budget is at least the amount passed by the user **after** the call is completed.

Reply:
fixed. resolving.